core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/util/DatabaseUtil.java

```diff
@@ -571,71 +570,116 @@ public static Connection getDBConnection(DataSource dataSource) throws SQLExcept
     }
 
     public static void closeConnection(Connection dbConnection) {
-
-        if (dbConnection != null) {
-            try {
-                dbConnection.close();
-            } catch (SQLException e) {
-                log.error("Database error. Could not close statement. Continuing with others. - " + e.getMessage(), e);
-            }
-        }
+        close(dbConnection);
     }
 
-    private static void closeResultSet(ResultSet rs) {
+    // Although using generics is a bit more cryptic to the uninitiated, this method ensures type safety
+    // and flexibility without the horrible cast conversions in
+    // favor of the usage of interfaces.
+    // Only java.sql types can be used as parameters.
+    //
+    // Only  java.sql object types intersect both Wrapper and AutoCloseable interfaces
+    // Types that are allowed as arguments: CachedRowSet, CallableStatement, Connection, DatabaseMetaData, DataSource,
+    // FilteredRowSet, JdbcRowSet, JoinRowSet, ParameterMetaData,
+    // PreparedStatement, ResultSet, ResultSetMetaData, RowSet, RowSetMetaData, Statement, SyncResolver, WebRowSet
+    private static <AutoClosableWrapper extends AutoCloseable & Wrapper> void close(AutoClosableWrapper dbObject) {
+
+        if (dbObject != null) {
 
-        if (rs != null) {
             try {
-                rs.close();
+                dbObject.close();
+            } catch (SQLRecoverableException ex) {
+                handleSQLRecoverableException(dbObject, ex, true);
             } catch (SQLException e) {
-                log.error("Database error. Could not close result set  - " + e.getMessage(), e);
+                log.error("Database error. Could not close statement. Continuing with others. - " + e.getMessage(), e);
+            } catch (Exception e) {
+                log.error("Generic error occurred during close operation" + e.getMessage(), e);
             }
         }
-
     }
 
-    private static void closeStatement(PreparedStatement preparedStatement) {
-
-        if (preparedStatement != null) {
-            try {
-                preparedStatement.close();
-            } catch (SQLException e) {
-                log.error("Database error. Could not close statement. Continuing with others. - " + e.getMessage(), e);
+    private static void handleSQLRecoverableException(AutoCloseable dbObject, SQLRecoverableException ex, boolean retry){
```

Comment:
Great stuff.  Maybe we can refactor line 72 in closeDatabasePoolConnection() to use this as well since DataSource in JDBC 7 is also an AutoCloseable object?